tests/xdc/stream_based_replication_test.go

```diff
@@ -0,0 +1,379 @@
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/common/primitives"
+	test "go.temporal.io/server/common/testing"
+	"go.temporal.io/server/service/history/replication"
+	"go.temporal.io/server/tests"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+type (
+	streamBasedReplicationTestSuite struct {
+		xdcBaseSuite
+		controller    *gomock.Controller
+		namespaceName string
+		namespaceID   string
+		serializer    serialization.Serializer
+		generator     test.Generator
+	}
+)
+
+func TestStreamBasedReplicationTestSuite(t *testing.T) {
+	suite.Run(t, new(streamBasedReplicationTestSuite))
+}
+
+func (s *streamBasedReplicationTestSuite) SetupSuite() {
+	s.controller = gomock.NewController(s.T())
+	s.dynamicConfigOverrides = map[dynamicconfig.Key]interface{}{
+		dynamicconfig.EnableReplicationStream:       true,
+		dynamicconfig.EnableEagerNamespaceRefresher: true,
+		dynamicconfig.EnableReplicationTaskBatching: true,
+	}
+	s.logger = log.NewNoopLogger()
+	s.serializer = serialization.NewSerializer()
+	s.setupSuite(
+		[]string{
+			"active",
+			"standby",
+		},
+		tests.WithFxOptionsForService(primitives.AllServices,
+			fx.Decorate(
+				func() config.DCRedirectionPolicy {
+					return config.DCRedirectionPolicy{Policy: "noop"}
+				},
+			),
+		),
+	)
+	ctx := context.Background()
+	s.namespaceName = "replication-test"
+	s.cluster1.GetFrontendClient().RegisterNamespace(ctx, &workflowservice.RegisterNamespaceRequest{
+		Namespace: s.namespaceName,
+		Clusters:  s.clusterReplicationConfig(),
+		// The first cluster is the active cluster.
+		ActiveClusterName: s.clusterNames[0],
+		// Needed so that the namespace is replicated.
+		IsGlobalNamespace: true,
+		// This is a required parameter.
+		WorkflowExecutionRetentionPeriod: durationpb.New(time.Hour * 24),
+	})
+	err := s.waitUntilNamespaceReplicated(ctx, s.namespaceName)
+	s.Require().NoError(err)
+
+	nsRes, _ := s.cluster1.GetFrontendClient().DescribeNamespace(ctx, &workflowservice.DescribeNamespaceRequest{
+		Namespace: s.namespaceName,
+	})
+
+	s.namespaceID = nsRes.NamespaceInfo.GetId()
+	s.generator = test.InitializeHistoryEventGenerator("namespace", "ns-id", 1)
+}
+
+func (s *streamBasedReplicationTestSuite) TearDownSuite() {
+	if s.generator != nil {
+		s.generator.Reset()
+	}
+	s.controller.Finish()
+	s.tearDownSuite()
+}
+
+func (s *streamBasedReplicationTestSuite) SetupTest() {
+	s.setupTest()
+	// Register a namespace.
+
+}
+
+func (s *streamBasedReplicationTestSuite) TestReplicateHistoryEvents_ForceReplicationScenario() {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, testTimeout)
+	defer cancel()
+
+	// let's import some events into cluster 1
+	historyClient1 := s.cluster1.GetHistoryClient()
+	executions := s.importTestEvents(historyClient1, namespace.Name(s.namespaceName), namespace.ID(s.namespaceID), []int64{3, 13, 2, 202, 302, 402, 602, 502, 802, 1002, 902, 702, 1102})
```

Comment:
just for my understanding.  is versions `failover version`? it doesn't need to guarantee to be monotonic increasing? what's the significance of the versions to do with the test?

Reply:
Yes, `versions` are failover versions. It need to guarantee to be monotonic increasing within single workflow history. Here I am importing different version of events into different workflows.